### PR TITLE
Fix planner bug that initplan may be removed mistakenly

### PR DESCRIPTION
**Pull request by hsyuan**
The old implementation of function `remove_unused_initplans` is buggy, we can
come up with multiple ways to break it, because it doesn't utilize postgres
plan/expression tree walker correctly. Rewrite the whole function to resolve
found and potential bugs.

### CODE REVIEW

File: src/backend/cdb/cdbmutate.c [hunk 1]
```diff
@@ -3019,105 +2922,49 @@ remove_unused_initplans_helper(Plan *plan, Bitmapset **usedParams, Bitmapset *rt
...
 {
-	if (NIL == rtable)
-	{
-		return NULL;
-	}
+	ParamWalkerContext context;
+	int num_subplans = list_length(root->glob->subplans);
 
-	Bitmapset *scanrelids = referenced_rtes(plan);
+	if (num_subplans == 0)
```
Comment (hlinnaka) [inline src/backend/cdb/cdbmutate.c]: This could be simplified to just `if (!root->glob->subplans)`

File: src/backend/cdb/cdbmutate.c [hunk 2]
```diff
@@ -3019,105 +2922,49 @@ remove_unused_initplans_helper(Plan *plan, Bitmapset **usedParams, Bitmapset *rt
...
+	context.scanrelids = NULL;
 
-	ListCell *lc = NULL;
-	int rteid = 0;
-	foreach (lc, rtable)
-	{
-		RangeTblEntry *rte = (RangeTblEntry *) lfirst(lc);
-		rteid ++;
+	/* get param ids from main plan and subplans */
+	param_walker((Node *) top_plan, &context);
```
Comment (hlinnaka) [inline src/backend/cdb/cdbmutate.c]: Let's improve the comments while we're at it. I'd suggest something like:

  Collect param ids of all the Params that are referenced in the Plan, and the IDs of all the range table entries that are referenced in the Plan.

File: src/backend/cdb/cdbmutate.c [hunk 3]
```diff
@@ -3019,105 +2922,49 @@ remove_unused_initplans_helper(Plan *plan, Bitmapset **usedParams, Bitmapset *rt
...
 
-		if (!bms_is_member(rteid, scanrelids))
-		{
-			// rte not referenced in the plan
-			continue;
-		}
+	/* get param ids that are referenced in RTEs */
+	rte_params_walker(root->glob->finalrtable, &context);
```
Comment (hlinnaka) [inline src/backend/cdb/cdbmutate.c]: And here:

Now that we know which range table entries are referenced in the plan, also collect Params from those range table entries.

File: src/backend/cdb/cdbmutate.c [hunk 4]
```diff
@@ -2787,220 +2787,123 @@ cdbhash_const_list(List *plConsts, int iSegments)
...
 
-/*
- * Remove unused initplans. This may happen if an initplan has been created, but
- * its corresponding param has been removed because of contradiction detection
- */
+/* retrieve param ids that are referenced in RTEs */
 static void
-remove_unused_initplans_helper(Plan *plan, Bitmapset **usedParams, Bitmapset *rte_params,
-							   PlannerInfo *root)
+rte_params_walker(List *rtable, ParamWalkerContext *context)
```
Comment (hlinnaka) [inline src/backend/cdb/cdbmutate.c]: This is very similar to range_table_walker(). I think I see why we can't just use range_table_walker() here, but would be good to have a comment on that.

**Comment by hsyuan** [comment]
Pushed to master, thanks for reviewing.

**Pull request merged**